tests/suites/test_suite_ssl.function

```diff
@@ -339,6 +341,264 @@ int mbedtls_mock_tcp_recv_nb( void *ctx, unsigned char *buf, size_t len )
+ * MBEDTLS_SSL_IS_CLIENT.
+ *
+ * \retval  0 on success, otherwise error code.
+ */
+int mbedtls_endpoint_init( mbedtls_endpoint *ep, int endpoint_type )
+{
+    int ret = -1;
+
+    if( ep == NULL )
+    {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+    }
+
+    ep->name = ( endpoint_type == MBEDTLS_SSL_IS_SERVER ) ? "Server" : "Client";
+
+    mbedtls_ssl_init( EP(ssl) );
+    mbedtls_ssl_config_init( EP(conf) );
+    mbedtls_ctr_drbg_init( EP(ctr_drbg) );
+    mbedtls_ssl_conf_rng( EP(conf),
+        mbedtls_ctr_drbg_random,
+        EP(ctr_drbg) );
+    mbedtls_entropy_init( EP(entropy) );
+    mbedtls_mock_socket_init( EP(socket) );
+
+    ret = mbedtls_ctr_drbg_seed( EP( ctr_drbg ), mbedtls_entropy_func,
+                    EP( entropy ), (const unsigned char*) ( ep->name ),
+                    strlen( ep->name ) );
+    TEST_ASSERT( ret == 0 );
+
+    /* Non-blocking callbacks without timeout */
+    mbedtls_ssl_set_bio( EP(ssl), EP(socket),
+        mbedtls_mock_tcp_send_nb,
+        mbedtls_mock_tcp_recv_nb,
+        NULL );
+
+    ret = mbedtls_ssl_setup( EP(ssl), EP(conf) );
+    TEST_ASSERT( ret == 0 );
+
+    ret = mbedtls_ssl_config_defaults( EP(conf), endpoint_type,
+                                        MBEDTLS_SSL_TRANSPORT_STREAM,
+                                        MBEDTLS_SSL_PRESET_DEFAULT );
+    TEST_ASSERT( ret == 0 );
+
+    ret = mbedtls_endpoint_certificate_init( ep );
+    TEST_ASSERT( ret == 0 );
+
+exit:
+    return ret;
+}
+
+/*
+ * Deinitializes certificates from endpoint represented by \p ep.
+ */
+void mbedtls_endpoint_certificate_free( mbedtls_endpoint *ep )
+{
+    mbedtls_endpoint_certificate *cert = &(ep->cert);
+    mbedtls_x509_crt_free( CERT(ca_cert) );
+    mbedtls_x509_crt_free( CERT(cert) );
+    mbedtls_x509_crt_free( CERT(cert2) );
+    mbedtls_pk_free( CERT(pkey) );
+    mbedtls_pk_free( CERT(pkey2) );
+}
+
+/*
+ * Deinitializes endpoint represented by \p ep.
+ */
+void mbedtls_endpoint_free( mbedtls_endpoint *ep )
+{
+    mbedtls_endpoint_certificate_free( ep );
+
+    mbedtls_ssl_free( EP(ssl) );
+    mbedtls_ssl_config_free( EP(conf) );
+    mbedtls_ctr_drbg_free( EP(ctr_drbg) );
+    mbedtls_entropy_free( EP(entropy) );
+    mbedtls_mock_socket_close( EP(socket) );
+}
+
+/*
+ * This function moves ssl handshake from \p ssl to prescribed \p state.
+ * /p second_ssl is used as second endpoint and their sockets have to be
+ * connected before calling this function.
+ *
+ * \retval  0 on success, otherwise error code.
+ */
+int mbedtls_move_handshake_to_state( mbedtls_ssl_context *ssl,
+                                     mbedtls_ssl_context *second_ssl,
+                                     int state )
+{
+    enum { BUFFSIZE = 1024 };
+    int ret = 0;
+
+    if( ssl == NULL || second_ssl == NULL )
+    {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+    }
+
+    /* Perform communication via connected sockets */
+    while( ssl->state != state )
```

Comment:
I'm concerned that this may cause a hang or infinite loop during testing if we can not get to the desired state. Please add a safety escape hatch so that if, say, we go around this loop 1000 times, we fail the test explicitly as it should never take 1000 steps to get to the state.